### PR TITLE
[2.8.5] Readded streaming music support.

### PR DESCRIPTION
**Pull request by raidensnake**
A redo of https://github.com/AttorneyOnline/AO2-Client/pull/338 to reimplement https://github.com/AttorneyOnline/AO2-Client/pull/134
Should resolve https://github.com/AttorneyOnline/AO2-Client/issues/336

### CODE REVIEW

**Comment by Crystalwarrior** [comment]
Wait is this seriously that simple
Fix the checks tho lel

**Review by Crystalwarrior — APPROVED** [review]

**Comment by raidensnake** [comment]
I hope I'm wring however it doesn't seem like macos can support streaming due to missing the BASS_StreamCreateURL function in it's library. Would like someone to check that as well.

**Review by stonedDiscord — APPROVED** [review]
yeah that's about how this bass streaming thing works.
No idea why it doesn't work on mac, but that is a bass problem, not ours

File: src/aomusicplayer.cpp [hunk 1]
```diff
@@ -32,6 +32,11 @@ void AOMusicPlayer::play(QString p_song, int channel, bool loop,
   else
     newstream = BASS_StreamCreateFile(FALSE, f_path.utf16(), 0, 0, flags);
 
+  if (f_path.startsWith("http"))
+    newstream = BASS_StreamCreateURL(f_path.toStdWString().c_str(), 0, BASS_STREAM_AUTOFREE | BASS_UNICODE | BASS_ASYNCFILE, NULL, NULL);
+  else
+	newstream = BASS_StreamCreateFile(FALSE, f_path.utf16(), 0, 0, flags);
```
Comment (oldmud0) [inline src/aomusicplayer.cpp:38]:
```suggestion
    newstream = BASS_StreamCreateFile(FALSE, f_path.utf16(), 0, 0, flags); 
```

File: src/aomusicplayer.cpp [hunk 2]
```diff
@@ -32,6 +32,11 @@ void AOMusicPlayer::play(QString p_song, int channel, bool loop,
   else
     newstream = BASS_StreamCreateFile(FALSE, f_path.utf16(), 0, 0, flags);
 
+  if (f_path.startsWith("http"))
+    newstream = BASS_StreamCreateURL(f_path.toStdWString().c_str(), 0, BASS_STREAM_AUTOFREE | BASS_UNICODE | BASS_ASYNCFILE, NULL, NULL);
+  else
+    newstream = BASS_StreamCreateFile(FALSE, f_path.utf16(), 0, 0, flags); 
+
```
Comment (oldmud0) [inline src/aomusicplayer.cpp:39]: Ah, gosh. Why did I push a style commit. This is wrong. The stream is being made twice.

File: src/aomusicplayer.cpp [hunk 3]
```diff
@@ -32,6 +32,11 @@ void AOMusicPlayer::play(QString p_song, int channel, bool loop,
   else
     newstream = BASS_StreamCreateFile(FALSE, f_path.utf16(), 0, 0, flags);
 
+  if (f_path.startsWith("http"))
+    newstream = BASS_StreamCreateURL(f_path.toStdWString().c_str(), 0, BASS_STREAM_AUTOFREE | BASS_UNICODE | BASS_ASYNCFILE, NULL, NULL);
```
Comment (oldmud0) [inline src/aomusicplayer.cpp:36]:
```suggestion
    newstream = BASS_StreamCreateURL(f_path.utf16(), 0, BASS_STREAM_AUTOFREE | BASS_UNICODE | BASS_ASYNCFILE, NULL, NULL);
```

**Review by oldmud0 — CHANGES_REQUESTED** [review]

File: src/aomusicplayer.cpp (hunk 3 above)
Comment (raidensnake) [reply src/aomusicplayer.cpp:36]: doing that causes an invalid conversion error that can only be overridden by an -fpermissive flag in windows. Still broken on linux and macosx.

File: src/aomusicplayer.cpp (hunk 2 above)
Comment (raidensnake) [reply src/aomusicplayer.cpp:39]: I tried condensing it by trying an #elseif however it didn't work.

File: src/aomusicplayer.cpp (hunk 3 above)
Comment (raidensnake) [reply src/aomusicplayer.cpp:36]: just like @stonedDiscord said. The bass header file doesn't allow streaming to work on other platforms even though the library itself has the symbol definitions to support it.

**Comment by stonedDiscord** [comment]
i don't agree with @oldmud0 suggested change to utf16
the header file expects a WCHAR and we should give it a WCHAR.
this way we don't need to lift the compiler restriction that does type checking

**Comment by stonedDiscord** [comment]
and line 33 is redundant, you can use an "else if"

**Comment by oldmud0** [comment]
Interesting that `BASS_StreamCreateURL` takes in a WCHAR. I checked the header and it appears that `BASS_StreamCreateURL` is overloaded specifically for Windows to take in a WCHAR instead of a `const char *`.  We're able to make the conversion with `utf16()` for `StreamCreateFile` because it takes in a `const void *` - but here the type is enforced.

**Comment by raidensnake** [comment]
the big question is can we adjust the code in the bass header file to support multiple platforms? The library symbols definitions are present in both the linux and macosx versions.

**Comment by oldmud0** [comment]
> The big question is can we adjust the code in the bass header file to support multiple platforms? The library symbol definitions are present in both the linux and macosx versions.

It is already defined on all platforms:

https://github.com/AttorneyOnline/AO2-Client/blob/b155ab8488cdd280d1b84ceaf0cd9fbff671a357/include/bass.h#L1098-L1100

**Comment by raidensnake** [comment]
> > The big question is can we adjust the code in the bass header file to support multiple platforms? The library symbol definitions are present in both the linux and macosx versions.
> 
> It is already defined on all platforms:
> 
> https://github.com/AttorneyOnline/AO2-Client/blob/b155ab8488cdd280d1b84ceaf0cd9fbff671a357/include/bass.h#L1098-L1100

Question how could we adjust the code depending on the platform?

**Comment by raidensnake** [comment]
> and line 33 is redundant, you can use an "else if"

@stonedDiscord Just to let you know I tried this and it didn't run stream at all. It has to be on it's own seperate lines.

**Comment by stonedDiscord** [comment]
fuck

**Comment by in1tiate** [comment]
fairly certain this was reimplemented somewhere else

**Pull request closed without merging**